Enhance docs with GitBook (#398)
src-d/lookout: docs/lookout-sdk.md

# lookout-sdk Binary

_For the **lookout-sdk** library to develop new analyzers go to [**lookout-sdk**](https://github.com/src-d/lookout-sdk) repository._

`lookout-sdk` binary is a simplified version of the lookout server that works with a local git repository and does not need access to Github.

> You can think about `lookout-sdk` as a _curl-like_ tool to call an analyzer gRPC endpoints with a ReviewEvent or a PushEvent, from a local git repository, and send it to an analyzer without accessing GitHub at all. For convenience, `lookout-sdk` also exposes a **lookout DataService** backed by the same git repository.

Reviewer: ```suggestion
You can think about `lookout-sdk` as a _curl-like_ tool to call an analyzer gRPC endpoint with a `ReviewEvent` or a `PushEvent`, from a local git repository, and send it to an analyzer without accessing GitHub at all. For convenience, `lookout-sdk` also exposes a **lookout DataService** backed by the same git repository.
```

You can download the latest `lookout-sdk` from the [src-d/lookout releases page](https://github.com/src-d/lookout/releases).

This is the sequence diagram of the `ReviewEvent` made by `lookout-sdk review`. You can compare it with a regular **lookout** installation at the [**source{d} lookout architecture** documentation](architecture.md).

> ![sequence diagram](assets/lookout-sdk-seq-diagram.png)

Reviewer: As I commented before. I don't like that a diagram shows `DataService` as a component, when it's just an internal part of the `lookout-sdk` command. Showing it at the same level as `lookout-sdk` may lead to believe you need to run something _else_, and it's not clear what.

I would merge `DataServer` with `lookout-sdk` in this diagram.

Author: addressed by 3484af7

```bash
$ docker-compose up bblfshd
```

This will create the [bblfshd](https://github.com/bblfsh/bblfshd) container listening on `localhost:9432`.

## Usage

To perform a `NotifyReviewEvent` call to an analyzer and serve the **lookout DataService** endpoint, run:
```shell
$ lookout-sdk review
```

To perform a `NotifyPushEvent` call to an analyzer and serve the **lookout DataService** endpoint, run:
```shell
$ lookout-sdk push
```

In the next section, you will find a more detailed example considering the most usual options for running `lookout-sdk` against any analyzer from two given revisions.

## How Does It Work

If we look at this example history of a given local repository stored at `/somewhere/repo/path`:

```bash
$ git log --pretty=oneline --graph

*   d036524c463227524f4bbd7b207fb87bb8b89ee3 (HEAD -> master) Merge pull request #3
|\
| * 045a24828327ac35a28186f9b9b437adc3f7b7a3 (branch-b) message
| * 804cbd94869cb173494ce1de410f2b48674bc772 message
|/
*   9294ddb13cc7c8acd2db480c9e5c1396cd85e50a (branch-a) message
*   7f2ee64cd0a5891900cc368ae35e60a61c262060 message
*
...
```

And considering we have an analyzer listening at `ipv4://localhost:9999`

we could run:

> ```shell

Reviewer: Wrong command, `branch-b` is not a parent of `branch-a`

$ lookout-sdk review \
  --git-dir=/somewhere/repo/path \
  --from=branch-b \
  --to=branch-a \
  "ipv4://localhost:9999"
```

Doing so, `lookout-sdk` will:

1. start a gRPC **lookout DataService** endpoint backed by the repository stored at `/somewhere/repo/path`;
1. create a `ReviewEvent` with the changes between `branch-b` and `branch-a`;
1. send a gRPC `NotifyReviewEvent` call to your analyzer listening on `ipv4://localhost:9999` with the created `ReviewEvent`;
1. wait till the analyzer sends a response with the comments; the analyzer will be able to request file contents, file language or UASTs to the gRPC **lookout DataService** endpoint exposed by `lookout-sdk`;
1. once the analyzer sends the response, `lookout-sdk` will put it into the `STDOUT`, stop the **lookout DataService** and exit.

_For more options to run `lookout-sdk`, take a look into [**lookout-sdk Command Options**](#options)_

> - If analyzer gRPC address is omitted, it will be `ipv4://localhost:9930`,

Reviewer: I'd prefer to not use trailing `,`

> - if `--git-dir` is omitted, it will be used the current dir,

Reviewer: ```suggestion
- if `--git-dir` is omitted, the current dir will be used
```

- if `--from` is omitted, it will be `HEAD^`,
- if `--to` is omitted, it will be `HEAD`,

> - both `--from` and `--to` can be a tag name, branch name or the full SHA-1 object name.

Reviewer: Well... If we get picky, `HEAD` is not any of these 3, right?
What about
```suggestion
- both `--from` and `--to` can be any [git revision](https://git-scm.com/docs/gitrevisions#_specifying_revisions). For example a tag name, branch name or the full commit SHA-1.
```

> All the explained above for `lookout-sdk review`, calling `NotifyReviewEvent`, applies also to `NotifyPushEvent` when using `lookout-sdk push`.

Reviewer: ```suggestion
Everything explained above for `lookout-sdk review` calling `NotifyReviewEvent`, applies also to `NotifyPushEvent` when using `lookout-sdk push`.
```
